File: Kalorie/Kalorie/Core/UseCases/CreateMyCreatedMealUseCase.swift
//
//  CreateMyCreatedMealUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 20.08.2026.
//

import Foundation

protocol CreateMyCreatedMealUseCaseProtocol {
    func callAsFunction(name: String, ingredients: [MyCreatedMealIngredientDomain]) async throws -> MyCreatedMealDomain
}

struct CreateMyCreatedMealUseCase: CreateMyCreatedMealUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(name: String, ingredients: [MyCreatedMealIngredientDomain]) async throws -> MyCreatedMealDomain {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        if let error = MyCreatedMealValidation.validate(name: name, ingredients: ingredients) { throw error }
        let now = Date.now
        let meal = MyCreatedMealDomain(
            id: UUID().uuidString,
            name: name.trimmingCharacters(in: .whitespacesAndNewlines),
            ingredients: ingredients,
            createdAt: now,
            updatedAt: now
        )
        let dto = MyCreatedMealDTO(meal: meal)
        try await dataProvider.setAsync(dto, id: meal.id, in: Constants.Firestore.myCreatedMeals(userId: userId))
        return meal
    }
}

#if DEBUG
struct CreateMyCreatedMealUseCaseFake: CreateMyCreatedMealUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(name: String, ingredients: [MyCreatedMealIngredientDomain]) async throws -> MyCreatedMealDomain {
        if shouldThrow { throw URLError(.unknown) }
        return MyCreatedMealDomain(id: UUID().uuidString, name: name, ingredients: ingredients, createdAt: .now, updatedAt: .now)
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/CreateMyCreatedMealUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodPortionDomain
import antoni.kalorie.core.models.MyCreatedMealDomain
import antoni.kalorie.core.models.MyCreatedMealIngredientDomain
import antoni.kalorie.core.models.MyCreatedMealValidation
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.MyCreatedMealDTO
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants
import java.time.Instant
import java.util.UUID

interface CreateMyCreatedMealUseCaseProtocol {
    suspend operator fun invoke(
        name: String,
        ingredients: List<MyCreatedMealIngredientDomain>,
        portions: List<FoodPortionDomain> = emptyList(),
    ): MyCreatedMealDomain
}

class CreateMyCreatedMealUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : CreateMyCreatedMealUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(
        name: String,
        ingredients: List<MyCreatedMealIngredientDomain>,
        portions: List<FoodPortionDomain>,
    ): MyCreatedMealDomain {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        MyCreatedMealValidation.validate(name = name, ingredients = ingredients, portions = portions)?.let { throw it }
        val now = Instant.now()
        val meal = MyCreatedMealDomain(
            id = UUID.randomUUID().toString().uppercase(),
            name = name.trim(),
            ingredients = ingredients,
            createdAt = now,
            updatedAt = now,
            portions = portions,
        )
        dataProvider.setAsync(MyCreatedMealDTO(meal), id = meal.id, inCollection = Constants.Firestore.myCreatedMeals(userId))
        return meal
    }
}
